CocoaPods Extension Build Support

dmiluski opened the pull request:
## 🦷 Issue:
Customer is unable to build Maps 5.9, / Events 0.10.2 if installed via CocoaPods.
- [x] https://github.com/mapbox/mapbox-events-ios/issues/47 Unable to Build run in extension (CocoaPods)
- [x] https://github.com/mapbox/mobile-telemetry/issues/754 Internal Tracking

## Why
CocoaPods is more strict than carthage where if installed via carthage one would see a warning, if installed via CocoaPods, a developer would see an error and be unable to run. This was not caught as carthage powered example shows a warning rather than a build failure error.

In this case, MapboxMobileEvents is leveraging access to sharedApplication which is not available for NSExtensions.

There is nothing preventing the SDK/Extension from building/running.
> ld: warning: linking against a dylib which is not safe for use in application extensions: /Users/danemiluski/Library/Developer/Xcode/DerivedData/MapboxMobileEvents-gtwngrjytwknvkfqciqcsoxcqjvr/Build/Products/Debug-iphonesimulator/MapboxMobileEvents.framework/MapboxMobileEvents

To enable friendlier App Extension usage though, we can signify we support app extension API only via this setting, which requires we don't require accessing SharedApplication api.

The way to ensure this SDK plays friendly when embedded in an extension is to mark extension only APIs.

## 🥅  Goal:
Provide mechanism for developer to install/run 
Provide an extensions that can be referenced in the future to validate no errors are found on run

## 🗺️ Approach
1. For each UIApplication.sharedApplication, provide an approach that plays friendly with NSExtension or UIApplication
2. Provide example extension showing a warning free build
3. Mark MapboxMobileEvents to only use extension APIs providing breaking feedback if one is used.

## 👷‍♂️ Changes
+ Leverage MMEUIApplicationWrapper to provide variations of data sources based on whether running in an App/Extension
+ Leverage responds to selector to call if running in UIApplication
+ Tests

## 📓  Notes:

Upon Testing, I did notice that the NSUserDefault persistent keys did not transfer between extension/app.  Will follow up to see if an app group in the TestHost enables the sharing. Although it is possible to share UserDefault values between extension and app, it requires an application passing in a known App Group ID to the SDK for use. This will require further discussion on actionable approaches.

## 📱 Screenshots
Demo Today Extension including MKMapView as demonstration for the possibility of including a map/telemetry gathering in extension

| Before: Project Extensions Option | After: Project Extensions Option |
| --- | --- |
| <img width="680" alt="Screen Shot 2020-06-08 at 5 01 26 PM" src="https://user-images.githubusercontent.com/5083390/84091633-e1bc4280-a9a9-11ea-981b-2ab7aba69b1a.png"> |  <img width="756" alt="Screen Shot 2020-06-11 at 11 47 46 AM" src="https://user-images.githubusercontent.com/5083390/84427629-14e51880-abda-11ea-8d30-335ffc8e7764.png"> |

| Before: sharedApplication Errors | After: sharedApplication Errors |
| --- | --- |
| <img width="1680" alt="Screen Shot 2020-06-08 at 5 11 37 PM" src="https://user-images.githubusercontent.com/5083390/84092132-63f93680-a9ab-11ea-9f6b-4dd31bdcca61.png"> | <img width="276" alt="Screen Shot 2020-06-11 at 11 55 02 AM" src="https://user-images.githubusercontent.com/5083390/84427839-67263980-abda-11ea-8594-579089e53396.png"> |


| Extension Demo GUI |
| --- |
| <img width="559" alt="Screen Shot 2020-06-08 at 4 27 07 PM" src="https://user-images.githubusercontent.com/5083390/84089902-1974bb80-a9a5-11ea-8b64-b9a0b28c56aa.png"> |

dmiluski committed:
Extension Example: Today Widget

Goal:
Show that MapboxMobileEvents SDK is able to be built/run in an extension.

Approach:
Provide a Today Widget example loading MapboxMobileEvents SDK and displaying (demonstrated on iPhone SE)

Outcome:
Able to build/run without crashing, although there is a warning due to the use of UIApplication.shared application directly inside of the framework.

> d: warning: linking against a dylib which is not safe for use in application extensions: /Users/danemiluski/Library/Developer/Xcode/DerivedData/MapboxMobileEvents-gtwngrjytwknvkfqciqcsoxcqjvr/Build/Products/Debug-iphonesimulator/MapboxMobileEvents.framework/MapboxMobileEvents

This warning could be remedied by enabling Allow app extension API only. Unfortunately, this would alos prevent the use of UIApplication.shared which can prove useful to identify UIApplication’s state.

dmiluski committed:
iOS (App Extension) Freindliness Continued

Context:
https://github.com/mapbox/mapbox-events-ios/pull/205

Although the above PR allowed for the content scale values to be extracted while in an extension, it did not cover all cases.

MMEUIApplicationWrapper also leverages UIApplication.shared to example application state as well as handle background task communication.

Although this is unavailable in NSExtensions, this can be stubbed out for the time being until a longer term decision is made upon how to handle lengthy actions/collections in an extenion.

For further support:
https://developer.apple.com/library/archive/documentation/General/Conceptual/ExtensibilityPG/ExtensionScenarios.html#//apple_ref/doc/uid/TP40014214-CH21-SW1

alfwatt commented:
@dmiluski I checked out the branch and everything is building OK but when I try to run the Release target I don't see the app in the "Add Widgets" list, any suggestions on debugging?

dmiluski commented:
> @dmiluski I checked out the branch and everything is building OK but when I try to run the Release target I don't see the app in the "Add Widgets" list, any suggestions on debugging?

Try running the TestHost app first, then running the extension.

dmiluski committed:
[App Extensions] Allow App Extension API Only

Problem:
When running, if marked extension API only, there are several errors to compile when accessing UIApplication.shared application.

This is due to:
- Lack of background task support through shared application
- Lack of app state

Goal:
Provide SDK friendly way to extract app state from sharedApplication is available. Otherwise, fallback to a default value.

Disable background task kick offs for extensions for the time being.

Changes:
- Remove MMEUINavigation at it is not used
+ Mark Allow app extension API only (Will error if build if sharedApplication is used directly providing feedback)
+ Extend MMEUIApplicationWrapper to infer sharedApplication or extension behaviors based on extension or not

rclee commented:
Upon revisiting the [original ticket](https://github.com/mapbox/mapbox-events-ios/issues/47). I was looking back over this [article](https://miqu.me/blog/2016/11/28/app-extensions-xcode-and-cocoapods-omg/)...

>Since Cocoapods 1.1.0, they improved integration with App extensions, and the generated project will enable the flag `APPLICATION_EXTENSION_API_ONLY` for libraries linking against an extension target. This is correct, but then you’ll start seeing compilation errors that can be a bit puzzling. Bear in mind that the code will not compile even if you don’t use the offending API. The compiler just complains that there is code that uses unavailable API.

Which may be why some users are seeing this as a build error instead of a warning.

dmiluski commented:
Verified in example repo that before this change CocoaPods installed run fails, and after succeeds.
https://github.com/mapbox/twc-collab/issues/1179#issuecomment-642293701

dmiluski committed:
[User Defaults Sharing] Sharing UserDefaults between App/Extension

Goal:
Experiment with UserDefault sharing between App/Extension

Outcome:
Through experimentation, confirmed what I wanted to clarify with the team on behaviors.

I access a shared user default between App/Extensions a UserDefaults(suite:) needs to be instantiated with the shared AppGroup provided from the top level application which has been configured with App Group.

If the goal is to support sharing user default values between App/Extension we would need to request our developers set the shared app group and we leverage it to access the shared resource.

Experiment:
Setup App Group and hop between App/Extension reading/writing to see what is visible between the two.

Manual Tests:
Evaluated the behavior of the App and extension by reading and writing. Launching each one after the other to verify whether the other could see the data written by the other

- Mismatch suite to App Group (No data sharing)
- Matchin suite with AppGroup (Data can be read/write between the two)

dmiluski committed:
[Flakey Test] Comment out

Test is refactored in 1.0 branch which should provide non-flakey coverage

dmiluski commented on MMETestHost/ViewController.m:
```diff
@@ -0,0 +1,32 @@
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
```
I experienced some difficulty debugging in the App delegate so Included this ViewController class as a place to experiment with NSUserDefaults. Although it could be removed from this PR, it may prove useful in the future to experiment with data sharing potential.

dmiluski commented on MapboxMobileEvents.xcodeproj/project.pbxproj:
```diff
@@ -1445,6 +1581,7 @@
 		404807CE1EA186D8008A6D50 /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				APPLICATION_EXTENSION_API_ONLY = YES;
```
The majority of this project file change have to do with the addition of a Today Widget.

This particular change though is what catches on build time if we ever use apis that don't work in extensions. Which will prevent this from happening in the future.

rclee reviewed (approved):
LGTM, Thanks for putting this together Dane!

dmiluski committed:
[Extension Support] Default Telemetry Off

Goal:
Currently, there is a devveloper broken with their attempt at integrating Maps/Telemetry support with their App extension.

We must also consider the ramifications of supporting telemetry gathering in an extension in general as questions around shared data store of the AppGroup as well as Privacy/Consent values have not been fully worked through yet.

Approach:
To unblock sooner rather than later, offer a supporting solution with an SDK using App Extension Only apis, but disable telemetry gathering if in an extension until interfaces and privacy considered expectations are finalized.

Changes:
- Disable Telemetry Gathering in App Extensions for the time being

dmiluski commented on MapboxMobileEvents/Categories/NSUserDefaults+MMEConfiguration.m:
```diff
@@ -373,6 +373,13 @@ - (BOOL)mme_isCollectionEnabled {
+    // Currently storage between App/Extension (AppGroup) is not shared meaning some values such as
+    // privacy review consent are not shared between both. Default collection to OFF until shared support
+    // is implemented
+    if (NSBundle.mme_isExtension) {
```
@alfwatt , 
After potential concerns about privacy setting sharing between App/Extension, opted to disable by defaults telemetry gathering in extensions until we can further reach clarity on what should be shared, as well as responsibilities to the developer to enable sharing.

alfwatt commented on MMETestHost/ViewController.m:
```diff
@@ -0,0 +1,32 @@
+    // Do any additional setup after loading the view.
+}
+
+/*
```
Not sure where the style guide is but generally we don't commit commented out code

alfwatt commented on MapboxMobileEvents/MMECommonEventData.m:
```diff
@@ -77,7 +78,7 @@ + (NSString *)applicationState {
         return MMEApplicationStateExtension;
     } else {
-        switch (UIApplication.sharedApplication.applicationState) {
+        switch (MMEUIApplicationWrapper.new.applicationState) {
```
We will eventually need an application state for 'Extension' to put into location events here, so we can track telemetry coming from extensions: https://github.com/mapbox/trace-maker/issues/97#issuecomment-638330549

dmiluski replied:
Yes, that is the line just above.
https://github.com/mapbox/mapbox-events-ios/pull/212/files#diff-016c26dee06c6b7561b0eee370142e89R79

Because this is leveraging an existing interface
https://developer.apple.com/documentation/uikit/uiapplication/1623003-applicationstate?language=objc

`@property(nonatomic, readonly) UIApplicationState applicationState;`

Where applicationState is an integer
```objective-c
typedef enum UIApplicationState : NSInteger {
    ...
} UIApplicationState;
```

It was easier to do the inspection above with an isExtension check. Is there an approach you would prefer otherwise?

alfwatt reviewed (approved):
Some suggested changes

misterkerry reviewed (approved):
nice progress

dmiluski commented on MapboxMobileEvents/MMEUIApplicationWrapper.h:
```diff
@@ -1,14 +1,54 @@
+
+// MARK: -
+
 @protocol MMEUIApplicationWrapper <NSObject>
```
Although possible, re-naming class/protocol conventions could cause pain for integration of 1.0 Beta branch in the future leading to an overly manual process.

How would you feel about holding off on this until after that is merged. Then more ideal naming/oranizing of files folders could take place such as support for SwiftPM.

dmiluski commented on Tests/MMELocationManagerTests.m:
```diff
@@ -42,29 +42,30 @@ - (void)testLocationManagerStartsMonitoringRegions {
-    
-    XCTAssert(capturedGeofence.center.longitude != newGeofence.center.longitude);
-}
+// Flakey Tests (refactored in 1.0 branch)
```
Will do

dmiluski replied:
Updated

dmiluski committed:
PR Feedback:

- Remove commented out code
- Remove commented out/unused property
- Remove unneeded comments
- Remove header doc
